mrp_bom_nested_bom_generator/models/mrp_bom.py

```diff
@@ -0,0 +1,144 @@
+        :yield list(MRP product, component)
+        :
+        """
+        nestings = [
+            self._prepare_temp_nested_bom_item(),
+        ] + list(self.nested_bom_ids.sorted(lambda line: line.sequence))
+        for i in range(1, len(nestings)):
+            yield nestings[i - 1], nestings[i]
+
+    def unlink_existing_bom(self) -> bool:
+        """
+        Unlink unused BOMs
+        Archive used BOM
+        Unlink all components in current BOM
+        :return: bool
+        """
+        child_ids = self.child_ids
+        if not child_ids:
+            return False
+        mrp_bom_used_ids = (
+            self.env["mrp.production"]
+            .search([("bom_id", "in", child_ids.ids)])
+            .mapped("bom_id")
+        )
+        mrp_bom_unused_ids = child_ids.filtered(
+            lambda bom: bom.id not in mrp_bom_used_ids.ids
+        )
+        mrp_bom_used_ids.sudo().write({"active": False})
+        mrp_bom_unused_ids.sudo().unlink()
+        self.bom_line_ids.sudo().unlink()
+        return True
+
+    def _create_mrp_bom_record(self, nested_bom, lines):
+        """
+        Create mrp.bom record for product
+        :param models.Model nested_bom: mrp.nested.bom record
+        :param list lines: x2m “create” command
+        :return bool: bool
+        """
+        if not lines or not nested_bom:
+            return False
+        if not all(len(line) == 3 for line in lines):
+            return False
+        if not all(not bool(c) and isinstance(v, dict) and v for c, _, v in lines):
+            return False
+        self.env["mrp.bom"].create(
+            {
+                "parent_id": self.id,
+                "product_tmpl_id": nested_bom.product_tmpl_id.id,
+                "product_uom_id": nested_bom.uom_id.id,
+                "product_qty": nested_bom.product_qty,
+                "type": "normal",
+                "bom_line_ids": lines,
+            }
+        )
+        return True
+
+    def create_boms(self) -> None:
+        """
+        Create Nested BOMs and unlink/archive old BOMs
+        :return None
+        """
+        self.unlink_existing_bom()
+        for index, arg in enumerate(self.group_by_stage()):
+            nested_bom, component = arg
+            bom_lines = component._prepare_bom_lines(nested_bom)
+            key = "bom_product_template_attribute_value_ids"
+            line_ptavs = [val[key] for _, _, val in bom_lines]
+            if all(line_ptavs) or len(bom_lines) == 1:
+                if bool(index):
+                    self._create_mrp_bom_record(nested_bom, bom_lines)
+                else:
+                    self.update({"bom_line_ids": bom_lines})
+                continue
+            for line in bom_lines:
+                if bool(index):
+                    self._create_mrp_bom_record(nested_bom, [line])
+                else:
+                    self.update({"bom_line_ids": [line]})
+
+    def action_generate_nested_boms(self) -> bool:
```

Comment:
Which records are generated here? mrp.bom or mrp.nested.bom.line?

Reply:
mrp.bom

Reply:
Renamed to action_generate_mrp_boms

Reply:
Resolved